### PR TITLE
openai retrieval function agent

### PR DESCRIPTION
**Pull request by jerryjliu**
# Description
Refactored base openai agent

Added RetrievalOpenAIAgent 

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: docs/examples/vector_stores/SimpleIndexDemoMMR.ipynb
Comment (Disiok) [inline docs/examples/vector_stores/SimpleIndexDemoMMR.ipynb:1]: I think this got in by accident?

File: llama_index/agent/openai_agent.py [hunk 1]
```diff
@@ -141,20 +120,22 @@ async def achat(
 
         return Response(ai_message.content)
 
-    def _get_function_by_name(self, name: str) -> BaseTool:
-        name_to_tool = {tool.metadata.name: tool for tool in self._tools}
+    def _get_function_by_name(self, tools: List[BaseTool], name: str) -> BaseTool:
```
Comment (Disiok) [inline llama_index/agent/openai_agent.py:123]: nit: moves this out of class to be utils function.

File: llama_index/agent/openai_agent.py [hunk 2]
```diff
@@ -141,20 +120,22 @@ async def achat(
...
+        name_to_tool = {tool.metadata.name: tool for tool in tools}
         if name not in name_to_tool:
             raise ValueError(f"Tool with name {name} not found")
         return name_to_tool[name]
 
-    def _call_function(self, function_call: dict) -> FunctionMessage:
+    def _call_function(
```
Comment (Disiok) [inline llama_index/agent/openai_agent.py:129]: nit: moves this out of class to be utils function.

File: llama_index/agent/openai_agent.py [hunk 3]
```diff
@@ -174,3 +155,130 @@ async def _aquery(self, query_bundle: QueryBundle) -> RESPONSE_TYPE:
...
+    NOTE: this is a beta feature, function interfaces might change.
+
+    TODO: add a native OpenAI Tool Index.
+
+    """
+
+    def __init__(
+        self,
+        retriever: BaseRetriever,
+        node_to_tool_fn: Callable[[Node], BaseTool],
```
Comment (Disiok) [inline llama_index/agent/openai_agent.py:227]: same comment as last time (i.e. `RetrieverRouterQueryEngine`)

I think taking in retriever and `node_to_tool_fn` is fairly unintuitive. What I mean is that if I only had access to this code, and not the notebook, I would not be able to conceptualize how to use this.

Comment (Disiok) [reply llama_index/agent/openai_agent.py:227]: I wouldn't know how to construct a retriever over tools, nor how to construct a node_to_tool_fn.

**Review by Disiok — APPROVED** [review]

File: llama_index/agent/openai_agent.py (hunk 3 above)
Comment (jerryjliu) [reply llama_index/agent/openai_agent.py:227]: chatted offline, will leave ofr now but think about how to improve interface

**Pull request merged**